### PR TITLE
Menu controller

### PR DESCRIPTION
**Pull request by mikilovski**
Created menu controller which reads from an RDF/XML layout and returns a custom list to generate the UI facets menu.

### CODE REVIEW

**Comment by Layoric** [comment]
@mikilovski This is looking really good! Concise loop to walk the RDF file to produce a simeple DTO that represents the menu. A nice addition would be a JUnit test that can test a known input and output to make sure it's behaving as expected. This would also help with writing the JS client as others can look at the unit test to see what the output looks like when serialized into JSON.

File: src/main/java/au/gov/ga/hydroid/controller/MenuController.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+@RestController
+@RequestMapping("/menu")
+public class MenuController {
+
+   @RequestMapping(value = "/hydroid", method = {RequestMethod.GET})
+   public @ResponseBody ResponseEntity<List<MenuDTO>> enhance() {
```
Comment (Layoric) [inline src/main/java/au/gov/ga/hydroid/controller/MenuController.java]: `enhance` isn't probably the best name for the method, but I'm assuming this was left over from getting started with a new controller. Something like `generateMenu` or something more descriptive of the controller function might suit better.

**Comment by Layoric** [comment]
@mikilovski Ha, ignore me, I missed the test between the large RDF files. Nicely done :+1:

File: src/test/java/au/gov/ga/hydroid/controller/MenuControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+   MenuController menuController;
+
+   @Before
+   public void setup() {
+      MockitoAnnotations.initMocks(this);
+      mockMvc = MockMvcBuilders.standaloneSetup(menuController).build();
+   }
+
+   @Test
+   public void testHydroidMenu() {
```
Comment (Layoric) [inline src/test/java/au/gov/ga/hydroid/controller/MenuControllerTest.java]: Could we add an additional test here to check the structure coming back? Eg, that the RDF provided is producing the correct nesting of items etc?

**Comment by scelton** [comment]
Nice and clean, looking great! I just added a json file under test/resources so we can compare with the response and made a small change to the controller to cache the menu and sort its items by label. Only issue I see is that the 3rd level menu is not being generated, might need to double check the rdf.

**Comment by Layoric** [comment]
Testing

**Comment by mikilovski** [comment]
I've fixed up recursions and seems to be generating the correct menu now. Unit test is passing with actual and expected from the json file in test/resources. I will add an additional test to check the structure soon.

**Comment by scelton** [comment]
Looks good!

**Comment by Layoric** [comment]
@mikilovski LGTM, good stuff re passing test of whole menu :+1:

**Comment by Layoric** [comment]
Looking good. I did make a couple of small changes. There was a decent chunk of code duplication when walking the RDF so I split it out into a separate function for reuse. Wasn't happy with the naming, so feed free to improve it if you think of something.

I also split the unit tests as generally one test function should only test one thing. This lets you more easily diagnose what is wrong when a test fails as it will give you the name of the test. If you have single responsibility tests (unit), then this process is a lot less confusing when it comes time to find out what change broke the unit tests.

@mikilovski have a look at the changes and let me know what you think.

**Comment by mikilovski** [comment]
@Layoric Happy with changes, cheers.

**Pull request merged**